MSRuleClener logic for output data rule cleanup

todor-ivanov opened the pull request:
Fixes #9880 

#### Status
Ready

#### Description
Implements the logic for the MSRuleCleaner micro service. It is meant to remove Rucio rules that are no longer needed in the Workload Management system, such as:                                                                                       
  * block-level     rules created by WMAgent, against the origin RSE where data is getting produced                                                                                                           
  * container-level rules created by WMAgent, with a very generic RSE expression                                                                                                                            
  * block-level     rules created by MSTransferor, for input data that is no longer in the system                                                                                                                     
  * container-level rules created by MSTransferor, for input data that is no longer in the system                                                                                                             

The Current PR introduces the full construction for the service, how it is meant to deal with the different types of rules to clean and to account for their cleanup status. This PR deals only with Container and Block level rules created by the WMAgents. It works with workflows in status `[ 'announced', 'rejected', 'aborted-completed']`. Here follows the different types of workflows recognized and the relevant actions taken:
1. `(RequestType == Resubmission)`: 
   * Do not clean any rules for them, but send them directly to be archived
2. `(RequestStatus  ==  rejected | aborted-completed) && (ParentageResolved == Any)`: 
   * Clean rules only from WMAgents (container && block level); 
   * Send them for archival 
3. `(RequestStatus == announced) && (ParentageResolved == False)`: 
   * Skip cleaning and archival steps; 
   * Log; 
   * Wait for the next cycle.
4. `(RequestStatus == announced) && (ParentageResolved == True)`:  
   * Clean rules from both WMAgents (block && container level) and MSTransferror (block && container level); 
   * Send them for archival.


#### Is it backward compatible (if not, which system it affects?)
YES 

#### Related PRs
It needs two more configuration values for the Rucio accounts used by WMAgents and MSTransferror:

https://github.com/dmwm/deployment/pull/970

#### External dependencies / deployment changes
No

todor-ivanov committed:
MSRuleClener basic logic.

Rebuild pipelines && refine docstrings && introduce 'runMode'.

Simplify workflow representatation for MSRuleCleaner

Execute Pipelines.

Implement initial dispatch logic && Introduce cleanup counters per pipeline.

Properly Trigger archival pipe line

Fetch transferInfo from MSOutput.

Implement 'PlineMarkers' as a list && Add logic for setting the final 'IsClean' flag.

Properly count cleaned workflows && Set IsClean flag && Add MSRuleCleanerArchival exception.

Some more logic in _dispatchWorkflow() && Moving pipelines creation in __init__().

Build the proper lists of workflows to populate 'RulesToClean'.

Fix some comments.

Add the cleanup logic.

Suppress the not yet build pipelines.

Fix some comments.

Fixes followoing the first code review.

Fixes followoing the first code review.

Fixes followoing the first code review.

Fixes followoing the first code review.

Fixes followoing the first code review.

Separate MSRuleCleanerWorkflow datastruct in a file.

Clean some notes.

Fixes followoing the second code review.

Fixes followoing the second code review.

amaltaro reviewed (commented):
This is still a very much work in progress, so it's not clear to me what review you'd like to have at this stage (please let me know if you have any specific questions).

I do have a comment though regarding `MSRuleCleanerTemplate`. Do we need a special class to store a few key/value pairs? Is there any reason not to make it as simple as possible and just stick to a python dictionary object? Last but not least, it relies on the workflow description, which already goes through checking/sanitization, so I'd be in favor of dropping that check - that is very convoluted in MSOutputTemplate - and keep it simple.

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +225,296 @@ def execute(self, reqStatus):
+            self.logger.exception(msg, wflow['PlineMarkers'][-1], str(ex))
+        return wflow
+
+    def setPlineMarker(self, wflow, pName=None):
```
I have a few comments on this method.
Why do we need to mark (set) which pipeline is executing the workflow? For error reporting?
I also think this is an overcomplication that should be avoided (besides having a fragile implementation).

todor-ivanov replied:
This we discussed in a private call also. As mentioned above I am working on approving it now.

todor-ivanov replied:
Done

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +225,296 @@ def execute(self, reqStatus):
+        #       supposed to be called as a standalone function in a pipeline.
+        # NOTE: `all([]) == True`, ergo all the 'rejected' && 'aborted-completed' workflows
+        #       are also counted as properly cleaned and can trigger archival later
+        cleanStatus = all([wflow['CleanupStatus'][key]
```
In addition to, this would be more readable if written in multiple lines.
I also think this is somehow a fragile construction, where we need to ensure that all the necessary pipelines get reflected in the `CleanupStatus` structure.

What came to my mind was that, all the `CleanupStatus` key/value pairs should be updated according to the status (read pipelines that need to be executed) and before a workflow gets archived, we need to make sure all of those are `True`.

todor-ivanov replied:
This  was the initial construction and was working wrong. It was giving false results, because that way one reads all the flags but does not take into account that based on the conditions from the upper logic (like workflow status, type of rules to be cleaned, parentage resolved flag and so on) a different set of pipelines are working on the workflow. So not only the counters were reported wrong but also archival was triggered for uncleaned workflows.  
This comment of yours is 1 to 1 related to the previous one, so leaving this one open too, because of the ongoing work.

amaltaro replied:
Perhaps something else to consider here is, to make sure that if any pipeline hits a bad exception, that we still have that pipeline writing its name and status to this workflow structure. Otherwise we could have things going to archive because the archival pipeline isn't aware of a broken pipeline in the chain.
I fail to see if this is a legit concern or if the code has a protection for that though.

todor-ivanov replied:
It is a legit concern, Yes. And that's  why I keep all default values to `False`. I usually try to watch out for such behavior. So what can ever happen are the following two scenarios:
   * The pipeline is marked (by having the very first function called to be `stPlineMarkers()`), but fails somewhere down the road... then what ever exception is thrown will be caught as a 'General Error' by the top level caller - `_dispatcher()` and the archival call will still be triggered, since it is the next step in the `_dispatcher()` but the flags would have already be set to `False` and no archival will happen.
   * The pipeline fails completely and it cannot even set it's markers then the whole `_dispatcher()` blows up at line 323 (currently) when it tries to increase an un-existing counter with: 
```                   self.wfCounters['cleaned'][pline.name] += 1 ```

I have not found a third sequence of events so far.

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +225,296 @@ def execute(self, reqStatus):
+        wflow['IsClean'] = self._checkClean(wflow)
+        return wflow
+
+    def archive(self, wflow):
```
Aren't you missing the actual status transition here?

todor-ivanov replied:
This is to be done while working in a separate GH issue related to the archival only.

amaltaro replied:
Please make sure this is reflected in the PR description.

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +225,296 @@ def execute(self, reqStatus):
+            msg = "Not properly cleaned workflow: %s" % wflow['RequestName']
+            raise MSRuleCleanerArchival(msg)
+
+        # Check the available status transitions before we decide the final status
```
I'd add something like:
```
if check_allowed_transition(wflow['RequestStatus'], wflow['TargetStatus']) is False:
  BIG exception
```
otherwise, do the request transition (and make a log record)

todor-ivanov replied:
Keep that though - to be worked in the next GH issue.

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -3,22 +3,142 @@
+                               'plineMSTrBlock': False,
+                               'plineAgentCont': False,
+                               'plineAgentBlock': False}, dict),
+            ('TransferInfo', None, dict),
```
Probably a better idea to make it boolean and call it like `MSOutputTransferStatus` or such.

todor-ivanov replied:
Done

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +225,296 @@ def execute(self, reqStatus):
+        mgr = RequestHandler()
+        res = mgr.getdata(url, params=params, headers=headers, ckey=ckey(), cert=cert())
+        data = json.loads(res)['result'][0]
+        wflow['TransferInfo'] = data['transferDoc']
```
I made a comment up in the data structure. I see no need to store the whole (even though it's not big) transfer info data structure here. In the end, we only care about the final status of the MSOutput output data placement.

todor-ivanov replied:
Done

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +225,296 @@ def execute(self, reqStatus):
+        :return:        wflow
+        """
+        currPline = wflow['PlineMarkers'][-1]
+        # Find all the output placement rules crated by the agents
```
typo in `crated`

todor-ivanov replied:
Thanks

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +225,296 @@ def execute(self, reqStatus):
+        # Find all the output placement rules crated by the agents
+        # NOTE: Check MSOutput 'TransferStatus' first and assure it is 'done'
+        #       only then we proceed with building the respective list of rules
+        if wflow['TransferInfo'] is not None \
```
See other couple of comments made above for `TransferInfo` and for fetching rules.

todor-ivanov replied:
Done

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +225,296 @@ def execute(self, reqStatus):
+           and wflow['TransferInfo']['TransferStatus'] == 'done': # FIXME 'pennding' -> 'done'
+            for dataCont in wflow['OutputDatasets']:
+                if gran == 'container':
+                    wflow['RulesToClean'][currPline].extend(self.rucio.listDataRules(dataCont, account=rucioAcct))
```
I guess you want to let Rucio module exceptions to propagate upstream, right?

todor-ivanov replied:
Yes, indeed. The exception will be caught up in the `_dispatcher()`, where the `pline.run()` method is called and the the execution for this workflow will be terminated.

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +225,296 @@ def execute(self, reqStatus):
+            self.logger.debug(msg, currPline, wflow['RequestName'])
+
+        # Set the cleanup flag:
+        wflow['CleanupStatus'][currPline] = all(delResutls)
```
no rules deleted will set it to True. I guess I'd remake this logic. See comment above.

todor-ivanov replied:
Changes to the code here won't protect us from having positive value when a `no rule was deleted` event happens. If you refer to an `all([])` behavior of this aggregation - I just never expect and empty list to land here because of the explanation given to the previous comment. While in case we want to avoid a list full of positive values when no rule was deleted at all then the change must happen in [2].  
Actually during my tests i did not see a problem with workflows having no rules to be deleted... or I suppose I should expect to have a workflow full of rules that have been already deleted and we hit that Exception for all of them ... but in this case I would consider the workflow already clean. I hope I am not missing something, btw. I will be glad to hear what do you think?

[2] 
 https://github.com/dmwm/WMCore/blob/51e3e2f6975b92abbbd9bf7291cf7e8d8eabd2e3/src/python/WMCore/Services/Rucio/Rucio.py#L719-L720

amaltaro replied:
I guess `delResutls` will only be `[]` if there were no rules to be deleted. Thus cleanup flag should go to True indeed.
Otherwise, it will have Falses/Trues according to the status of the rule deletion, which the `all` builtin will properly evaluate. So I think we are good here.

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +225,296 @@ def execute(self, reqStatus):
+        if wflow['TransferInfo'] is not None \
+           and wflow['TransferInfo']['TransferStatus'] == 'done': # FIXME 'pennding' -> 'done'
+            if self.isRealMode:
+                for rule in wflow['RulesToClean'][currPline]:
```
How about something like:
```
successfulDeletion = True
for rule in wflow['RulesToClean'][currPline]:
   if fail to delete any rule:
      successfulDeletion = False
...
wflow['CleanupStatus'][currPline] = successfulDeletion
```
?

todor-ivanov replied:
Exactly the same logic is already implemented internally into the `rucio.deleteRule()` method. See [1]. No need to repeat here I believe. Will gain nothing. So I already count on the fact that I have no way to get anything else from inside that method but a `Bool` value i.e. it already suppresses any exception that my be thrown by the third parity Rucio client. 

[1]
https://github.com/dmwm/WMCore/blob/51e3e2f6975b92abbbd9bf7291cf7e8d8eabd2e3/src/python/WMCore/Services/Rucio/Rucio.py#L717-L724

amaltaro replied:
My comment was on how you make sure if you have successfully deleted all the rules or not (basically your logic with `delResult`). Just a different way to evaluate it though.

amaltaro reviewed (changes requested):
There you go!

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -3,22 +3,142 @@
- * block-level rules created by WMAgent, against the origin RSE where data is getting produced
- * container and block-level rules created by MSTransferor, for input data that is no longer in the system
+ * block-level     rules created by WMAgent, against the origin RSE where data is getting produced
+ * container-level rules created by WMAgent, against the origin RSE where data is getting produced
```
Done

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -3,22 +3,142 @@
 In addition to that, this MicroService is now also responsible for the workflow archival, which is the final status that
 workflows remain.
 """
+# NOTE:
```
No, this comment was  not intended to persist in the code. It was used here  just to serve some visualization purpose. Removing it now.

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -3,22 +3,142 @@
+from WMCore.ReqMgr.DataStructs import RequestStatus
 from Utils.EmailAlert import EmailAlert
+from Utils.Pipeline import Pipeline, Functor
+from Utils.Timers import timeFunction
```
removed

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -3,22 +3,142 @@
+        super(MSRuleCleanerException, self).__init__(self.myMessage)
+
+
+class MSRuleCleanerArchival(MSRuleCleanerException):
```
agreed on the general `MSRuleCleaner` Exception and it is removed now.
When it comes to the `myMessage` attribute it is referenced later in the code. I will give the reason why in the related comment bellow.

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -3,22 +3,142 @@
+        super(MSRuleCleanerArchival, self).__init__(self.myMessage)
+
+
+class MSRuleCleanerCleanup(MSRuleCleanerException):
```
Yes indeed, I was planning to use it to signal a failed cleanup action per workflow from inside the `cleanRuciorules` function so that the counters for this workflows are not updated in a way similar to the `MSRulecleanerArchival` Exception.... but... as we can see the implementation did not follow the original design. I am deleting it now.

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -3,22 +3,142 @@
+    def __init__(self, doc, **kwargs):
+        super(MSRuleCleanerWflow, self).__init__(**kwargs)
+
+        # Search for all the keys we need from the ReqManager workflow describtion
```
thanks

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -3,22 +3,142 @@
+                              'plineMSTrBlock': [],
+                              'plineAgentCont': [],
+                              'plineAgentBlock': []}, dict),
+            ('CleanupStatus', {'plineMSTrCont': False,
```
As we discussed it in a private call, we will be keeping the number of pipe lines for the moment, and I am now  following a different path when it comes to the creation of those keys here, so that  we avoid errors like typos etc.

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -38,10 +158,65 @@ def __init__(self, msConfig, logger=None):
+        # FIXME - make the proper PR to the config repo
+        self.msConfig.setdefault("rucioWmaAccount", "wma_test")
+        self.msConfig.setdefault("rucioMStrAccount", "wmcore_transferor")
+        self.msConfig.setdefault("runMode", "dryRun")
```
Agreed && Done.

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -38,10 +158,65 @@ def __init__(self, msConfig, logger=None):
         self.emailAlert = EmailAlert(self.msConfig)
 
+        # Building all the Pipelines:
+        # NOTE: The pipeline names need to overlap with the key names in the
```
That is fragile for sure. That is the reason why this comment was left here. But I agree a comment does not solve an issue. I am automating this process now. Will post in this section later, once I am done with this implementation.

todor-ivanov replied:
Done

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -38,10 +158,65 @@ def __init__(self, msConfig, logger=None):
+                                                 Functor(self.getRucioRules, 'container', self.msConfig['rucioWmaAccount']),
+                                                 Functor(self.cleanRucioRules)])
+        pName = 'plineAgentBlock'
+        self.plineAgentBlock = Pipeline(name=pName,
```
As  already mentioned  we are keeping the number of `plines` for the moment. We will eventually explore that option later.

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -38,10 +158,65 @@ def __init__(self, msConfig, logger=None):
+        self.mstrlines = [self.plineMSTrCont,
+                          self.plineMSTrBlock]
+
+        # Zero the 'cleaned' and 'archived' counters:
```
I do not see any need for a separate method, because these counters are initialized only once during the thread creation and never reset again until the end. I would rather change the comment saying:
`Initializing the 'cleaned' and 'archived' counters`

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +225,296 @@ def execute(self, reqStatus):
+        self.currThreadIdent = self.currThread.name
+        self.updateReportDict(summary, "thread_id", self.currThreadIdent)
+
+        msg = "{}: MSRuleCleaner is running in mode: {}.".format(self.currThreadIdent, self.mode)
```
Done

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +225,296 @@ def execute(self, reqStatus):
-            msg = "  retrieved %s requests. " % len(requestRecords)
-            msg += "Service set to process up to %s requests per cycle." % self.msConfig["limitRequestsPerCycle"]
+            requestRecords = {}
+            for status in reqStatus:
```
Yes, Way too big change in order to be covered here. If we decide to give this a try we better follow in a separate issue.

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +225,296 @@ def execute(self, reqStatus):
+            msg = "\n{}: Number of processed workflows: {}.".format(self.currThreadIdent, total_num_requests)
+            msg += "\n{}: Number of properly cleaned workflows: {}.".format(self.currThreadIdent, clean_num_requests)
+            msg += "\n{}: Number of archived workflows: {}.".format(self.currThreadIdent, archived_num_requests)
+            msg += "\n{}: Service set to process up to {} workflows per cycle.".format(self.currThreadIdent, self.msConfig["limitRequestsPerCycle"])
```
Done

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +225,296 @@ def execute(self, reqStatus):
-        self.updateReportDict(summary, "error", 42)
+        # Call _execute() and feed the relevant pipeline with the objects popped from requestRecords
+        try:
+            clean_num_requests, total_num_requests, archived_num_requests = self._execute(requestRecords)
```
I used that  name intentionally, because this is actually an auxiliary function to `execute()`. It is basically used to do the job for the top level execute function, but only this time it splits the work by workflow instead of working on bulk.

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +225,296 @@ def execute(self, reqStatus):
+    def _execute(self, reqRecords):
+        """
+        Executes the MSRuleCleaner pipelines based on the workflow status
+        :param reqList: A list of ReqManager records to work on
```
Done

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +225,296 @@ def execute(self, reqStatus):
+
+        # Call the workflow dispatcher:
+        for reqName, req in reqRecords.items():
+            wflow = MSRuleCleanerWflow(reqRecords.pop(reqName))
```
Done

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +225,296 @@ def execute(self, reqStatus):
+        """
+        # NOTE: The following dispatch logic is a subject to be changed at any time
+
+        if wflow['RequestType'] == 'Resubmission':
```
Agreed, since `_execute()` is taking care of splitting the work in smaller junks all filters applied should definitely go in here instead of the dispatcher. And actually it makes the `_dispatcherWflow()` method look much more structured.

Done.

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +225,296 @@ def execute(self, reqStatus):
+                    msg = "%s: General error from pipeline. Error: \n%s. "
+                    msg += "\nWill retry again in the next cycle."
+                    self.logger.exception(msg, pline.name, str(ex))
+                    break
```
I was just trying to spare some cycles here.... but yes if we agree that shortening the rules lifetime is a bigger priority this is the right thing to do.
Done

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +225,296 @@ def execute(self, reqStatus):
+                    break
+                if wflow['CleanupStatus'][pline.name]:
+                    self.wfCounters['cleaned'][pline.name] += 1
+        elif wflow['RequestStatus'] == 'announced' and wflow['ParentageResolved']:
```
Done, But... It may be also considered as a filter and I wondered if it would not be a better idea to move it in `_execute()` where I started applying the rest of the filters with moving the `Resubmission` up. In both cases we should take into account there is a default value for the 'ParentageResoved'  flag set to `True`  in the MSRuleCleaner object creation.  
We should decide this before. Tell me what do you think?

amaltaro replied:
It could be done over there too. Maybe even wrapping those 2 checks in an utilitarian method and only convert to MSRuleCleanerWflow object if it passes those constraints...

amaltaro replied:
Oh, and this last `elif` statement can likely be an `else` now, right?

todor-ivanov replied:
yes, it can

todor-ivanov replied:
ok ... thinking twice .. I like it more with an `elif`, because it states explicitly on which workflows it is about to be working on.... if we make it an `else`, than we make it too general and everything which does not suffice  the above conditions will fall under this category and will be worked on just like the proper `announced` with 'ParentageResolved'. In the current set of conditions above we are safe, because we expect only those to reach down here, but imagine we decide to put or take out something above later on and forget to take this int account here... then... voile a bug is stalking behind the corner :)

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +225,296 @@ def execute(self, reqStatus):
+        except MSRuleCleanerArchival as ex:
+            msg = "%s: Archival Error: %s. "
+            msg += "\nWill retry again in the next cycle."
+            self.logger.error(msg, wflow['PlineMarkers'][-1], ex.myMessage)
```
I tried this and the logs become really blurred. It adds alot of extra (multi-lined) info comming from the WMCore Exception, including fetched job error codes and so on, which are totally not relevant to our case, and it will pop out in every workflow for which archival is skipped. it makes the whole log pretty much unreadable. See the difference between [1] and [2].

[1]
```
2020-11-04 08:38:49,651:DEBUG:MSRuleCleaner:cleanRucioRules(): plineAgentCont: 'TransferStatus' is 'pending' or 'TransferInfo' is missing. Skip deleting any rules for: tivanov_TC_LHE_PFN_HG2011_Val_201029_112700_5697 
2020-11-04 08:38:49,682:DEBUG:MSRuleCleaner:cleanRucioRules(): plineAgentBlock: 'TransferStatus' is 'pending' or 'TransferInfo' is missing. Skip deleting any rules for: tivanov_TC_LHE_PFN_HG2011_Val_201029_112700_5697 
2020-11-04 08:38:49,703:ERROR:MSRuleCleaner:_dispatchWflow(): plineArchive: Archival Error: <@========== WMException Start ==========@>
Exception Class: MSRuleCleanerArchival
Message: MSRuleCleanerArchival: Not properly cleaned workflow: tivanov_TC_LHE_PFN_HG2011_Val_201029_112700_5697
	ModuleName : WMCore.MicroService.Unified.MSRuleCleaner
	MethodName : __init__
	ClassInstance : None
	FileName : /afs/cern.ch/user/t/tivanov/work/private/WMCoreDev.d/WMCore/src/python/WMCore/MicroService/Unified/MSRuleCleaner.py
	ClassName : None
	LineNumber : 41
	ErrorNr : 0

Traceback: 

<@---------- WMException End ----------@>. 
 ```

[2]
```
2020-11-04 08:40:36,497:DEBUG:MSRuleCleaner:cleanRucioRules(): plineAgentBlock: 'TransferStatus' is 'pending' or 'TransferInfo' is missing. Skip deleting any rules for: tivanov_TC_LHE_PFN_HG2011_Val_201029_112700_5697 
2020-11-04 08:40:36,518:ERROR:MSRuleCleaner:_dispatchWflow(): plineArchive: Archival Error: MSRuleCleanerArchival: Not properly cleaned workflow: tivanov_TC_LHE_PFN_HG2011_Val_201029_112700_5697. 
Will retry again in the next cycle.
```

amaltaro replied:
I see. How about calling this method then?
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WMException.py#L161

todor-ivanov replied:
It does the job. Thanks!

amaltaro commented:
test this please

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -3,22 +3,142 @@
+            ('TransferInfo', None, dict),
+            ('TargetStatus', None, (str, unicode)),
+            ('ParentageResolved', True, bool),
+            ('PlineMarkers', [None], list),
```
Done

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +225,296 @@ def execute(self, reqStatus):
+        if wflow['RequestType'] == 'Resubmission':
+            pass
+
+        # Clean:
```
it does the job. Thanks!

amaltaro commented:
@todor-ivanov since I will be a bit late with my review, please update a few things that we have already discussed, such as:
* a proper description and summary of the behaviour in the first comment of this PR
* separate the `MSRuleCleanerWflow` class in a different module under `WMCore/MicroService/DataStructs`

I'll review it this afternoon and there might be further requests.

todor-ivanov commented:
Thanks @amaltaro, you already have the requested change and I just updated the PR description.

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +116,341 @@ def execute(self, reqStatus):
+            self.logger.info(msg, pline.name, self.wfCounters['cleaned'][pline.name])
+        archived_num_requests = self.wfCounters['archived']
+        msg = "Workflows archived: %d"
+        self.logger.info(msg, self.wfCounters['archived'])
```
This `msg` and the logger line could also be merged in a single line.

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +116,341 @@ def execute(self, reqStatus):
+        if wflow['RequestType'] == 'Resubmission':
+            wflow['ForceArchive'] = True
+            msg = "Skipping cleanup step for workflow: %s - RequestType is %s."
+            msg += "\nWill try to archive it directly."
```
How about we keep all this log in the same line? It's usually helpful when scanning the service logs and debugging it.

todor-ivanov replied:
done

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +116,341 @@ def execute(self, reqStatus):
+                    pline.run(wflow)
+                except Exception as ex:
+                    msg = "%s: General error from pipeline. Error: \n%s. "
+                    msg += "\nWill retry again in the next cycle."
```
I'd suggest to keep this one in the same line as well. Shouldn't it include the workflow name too?

todor-ivanov replied:
Totally agree when it comes to the Workflow name. And disagree when it comes to including the General exception and the text in the same line. As you mentioned in the preovious comments it would be good to keep those on the same line, and i did so when it comes to a well defined messages, but when doing `str(ex)` it is basically a stack dump and it is multi line message anyway and not separating it on a new line just smears the main message with the stack dump.

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +116,341 @@ def execute(self, reqStatus):
+            # NOTE: We skip workflows which are not having 'ParentageResolved'
+            #       flag, but we still need some proper logging for these.
+            msg = "Skipping workflow: %s - 'ParentageResolved' flag set to false."
+            msg += "\nWill retry again in the next cycle."
```
Please keep all this message in the same line too.

todor-ivanov replied:
Done

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +116,341 @@ def execute(self, reqStatus):
+                    pline.run(wflow)
+                except Exception as ex:
+                    msg = "%s: General error from pipeline. Error: \n%s. "
+                    msg += "\nWill retry again in the next cycle."
```
Same comment here. Having the workflow in the error message is important too.

todor-ivanov replied:
Workflow name was  added to the message.

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +116,341 @@ def execute(self, reqStatus):
+            # We shouldn't be here:
+            msg = "Skipping workflow: %s - "
+            msg += "Does not fall under any of the defined categories."
+            self.logger.info(msg, wflow['RequestName'])
```
Shouldn't it be an ERROR logging level?

todor-ivanov replied:
Agreed.
Done

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +116,341 @@ def execute(self, reqStatus):
+            msg = "%s General error from pipeline. Error: \n%s. "
+            msg += "\nWill retry again in the next cycle."
+            self.logger.exception(msg, wflow['PlineMarkers'][-1], str(ex))
+        return wflow
```
no need to return `wflow`, right?

todor-ivanov replied:
correct

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +116,341 @@ def execute(self, reqStatus):
+            msg += "Does not fall under any of the defined categories."
+            self.logger.info(msg, wflow['RequestName'])
+
+        # Archive:
```
First thought that came to my mind was: it deserves a separate method. Minor though, so we might come back to it in the future...

todor-ivanov replied:
This thought you may need to explain to me in more details. But I suggest to do it in one of the next calls we have on the topic.

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +116,341 @@ def execute(self, reqStatus):
+        except MSRuleCleanerArchival as ex:
+            msg = "%s: Archival Error: %s. "
+            msg += "\nWill retry again in the next cycle."
+            self.logger.error(msg, wflow['PlineMarkers'][-1], ex.message())
```
Same comment, should we add the workflow name in the error message and write it all in the same line? When using the logger object, breaking these in multiple lines make things look bad (besides losing the timestamp from a possible grep command).

todor-ivanov replied:
The workflow name is already in the message coming from the `archive()` method, which puts it directly into the exception message. See the log messages related to the accumulated action history of a non-archived workflow [1]. 
The Redundant new line was removed.

Done

[1]
```
2020-11-06 15:51:40,397:DEBUG:MSRuleCleaner:cleanRucioRules(): plineAgentCont: 'TransferStatus' is 'pending' or 'TransferInfo' is missing. Skip deleting any rules for: tivanov_TaskChain_Prod_HG2011_Val_201029_112740_7125 
2020-11-06 15:51:40,427:DEBUG:MSRuleCleaner:cleanRucioRules(): plineAgentBlock: 'TransferStatus' is 'pending' or 'TransferInfo' is missing. Skip deleting any rules for: tivanov_TaskChain_Prod_HG2011_Val_201029_112740_7125 
2020-11-06 15:51:40,449:ERROR:MSRuleCleaner:_dispatchWflow(): plineArchive: Archival Error: MSRuleCleanerArchival: Not properly cleaned workflow: tivanov_TaskChain_Prod_HG2011_Val_201029_112740_7125. Will retry again in the next cycle.
```

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -3,22 +3,42 @@
+    Archival Exception Class for MSRuleCleaner Module in WMCore MicroServices
+    """
+    def __init__(self, message):
+        self.myMessage = "MSRuleCleanerArchival: %s" % message
```
`myMessage` does not need to be an object attribute.

todor-ivanov replied:
Done

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +116,341 @@ def execute(self, reqStatus):
+            self.logger.exception(msg, wflow['PlineMarkers'][-1], str(ex))
+        return wflow
+
+    def setPlineMarker(self, wflow, pName=None):
```
I guess you want to make `pName` a mandatory argument?

todor-ivanov replied:
Done

amaltaro commented on src/python/WMCore/MicroService/DataStructs/MSRuleCleanerWorkflow.py:
```diff
@@ -0,0 +1,80 @@
+            "TransferDone": False  # information - returned by the MSOutput REST call.
+            'TargetStatus': 'normal-archived' || 'rejected-achived' || 'aborted-archived',
+            'ParentageResolved': Bool,
+            'PlineMarkers': [None],
```
This line needs to be updated.

todor-ivanov replied:
Done

amaltaro commented on src/python/WMCore/MicroService/DataStructs/MSRuleCleanerWorkflow.py:
```diff
@@ -0,0 +1,80 @@
+            ('TransferDone', False, bool),
+            ('TargetStatus', None, (str, unicode)),
+            ('ParentageResolved', True, bool),
+            ('PlineMarkers', None, list),
```
Didn't we discuss that its default value would be an empty list?

todor-ivanov replied:
I remember I was keeping this because I wanted to have a way to distinguish between the cases when 'It is the very first pLine executed on that workflow' and  'No one have done any job on the workflow at all'. Since I have tested it with `None` value I'd rather keep it as it is. It will require a new set of workflows injected into the system so that I can test with a `[]` default.

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +116,341 @@ def execute(self, reqStatus):
+        #       end of the 'PlineMarkers' list
+
+        # First get rid of the default:
+        if not wflow['PlineMarkers']:
```
Depending on what happens to the default value, these 2 lines might be deleted.

todor-ivanov replied:
As already mentioned - I would keep it unchanged for the time being.

amaltaro replied:
You kept the `None` initial value, so these 2 lines are still meaningful. Marking it as resolved. Thanks

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +116,341 @@ def execute(self, reqStatus):
+
+        # Populate the list of flags to be used later:
+        if pName not in wflow['RulesToClean']:
+            if pName in [pline.name for pline in self.cleanuplines]:
```
Just an observation (same for the block below):
```
[pline.name for pline in self.cleanuplines]
```
is an static code, even though it takes some CPU cycles for every workflow and pileup to be executed. Likely something that's worth changing.

todor-ivanov replied:
I am open for suggestions. This was the shortest, cleanest and cheapest (as it comes to CPU cycles and lines of code) I could come up with. And I also tested all the use cases I could cover in the same format (as for the check a line bellow) so that it is crystal clear just from reading the code which are the use cases covered in these checks.

amaltaro replied:
What I had in mind was something like
```
self.cleanupPipeNames = [pline.name for pline in self.cleanuplines]
```

somewhere upstream, like `__init__` or `execute`. Then here it would be just 
```
if pName in self.cleanupPipeNames:
```

Python interpreter might be smart enough to optimize the original line though, making no use of the enhancement I'm suggesting. If you feel like not changing this code anymore as well, that's fine either.

todor-ivanov replied:
Done

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +116,341 @@ def execute(self, reqStatus):
+        # NOTE: `all([]) == True`, ergo all the 'rejected' && 'aborted-completed' workflows
+        #       are also counted as properly cleaned and can trigger archival later
+
+        # Build a list of bool flags based on the mask of PlineMarkers
```
Please break this down into multiple lines to make it more readable.

todor-ivanov replied:
Again, I couldn't imagine a more readable way to express how this list of flags to be checked is build.

amaltaro replied:
It's really a matter of taste here... I tend to prefer constructions like:
```
cleanFlagsList = []
for key in wflow['PlineMarkers']:
    if key in wflow['CleanupStatus'].keys():
        cleanFlagsList.append(wflow['CleanupStatus'][key])
```

instead of multi-line list comprehensions. Anyhow, marking as resolved too.

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +116,341 @@ def execute(self, reqStatus):
+                wflow['CleanupStatus'][pName] = False
+        return wflow
+
+    def _checkClean(self, wflow):
```
I have a hard time picturing all the combinations here. It's also a method that can screw up with the workflow cleanup. So, I'd suggest to get a unit test exercising this logic. No need to run the whole microservice, but this method is very important IMO.

todor-ivanov replied:
So If you allow me I may try to visualize it like that:
We start from the fact that we apply a mask of named flags from one list (the list of pipelines that have worked on a workflow) over a list of `Bool` flags (The list of cleanup results). And we also need to take into account that the two lists may not overlap because of the fact that not every workflow will fall under the same logic of cleaning - some may have cleaned just from rules from the Agents other from all MStranferoror and Agents and so on. So the table of valid use cases is [1]. When the last one i consider a bug because it will produce wrong results. That's why we  have  the last condition in the code above.




[1]
```
|--------------+---------------------------+---------------------+---------------------------------------|
| Use cases    | Lists                     | Contents            | Example                               |
|--------------+---------------------------+---------------------+---------------------------------------|
| MasList      | Mask:                     | M1  M2  --          | [plMStrCont, plMStrBlock]             |
| shorter than | Flags:                    | F1  F2  F_extraFlag | [True,       True,        False ]     |
| FlagList     | Flags taken into account: | {F1, F2}            |                                       |
|              | Final Value:              | all([F1, F2])       | all([True, True]) = True              |
|              |                           |                     |                                       |
|--------------+---------------------------+---------------------+---------------------------------------|
| MaskList     | Mask:                     | --                  | []                                    |
| Empty        | Flags:                    | F1 F2 F3            | [True,          True,        False ]  |
|              | Flags taken into account: | --                  |                                       |
|              | Final Value:              | False               | False                                 |
|              |                           |                     |                                       |
|--------------+---------------------------+---------------------+---------------------------------------|
| MaskList     | Mask:                     | M1  M2  M3_extra    | [plMStrCont, plMStrBlock, plArchiver] |
| longer than  | Flags:                    | F1  F2              | [True,       True ]                   |
| Flaglist     | Flags taken into account: | {F1, F2}            |                                       |
|              | Final Value:              | all([F1, F2])       | all([True, True]) = True              |
|              |                           |                     |                                       |
|--------------+---------------------------+---------------------+---------------------------------------|
| FlagList     | Mask:                     | M1  M2              | [plMStrCont, plMStrBlock]             |
| empty        | Flags:                    | --  --              | []                                    |
|              | Flags taken into account: | {}                  |                                       |
|              | Final Value:              | all([])             | all([]) = True                        |
|              |                           |                     |                                       |
|--------------+---------------------------+---------------------+---------------------------------------|
```

amaltaro replied:
Thanks, let's see how the unit tests look like ;)

todor-ivanov replied:
Done - The unit tests are implemented and are referring this comment for visualization purposes of the test itself.

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +116,341 @@ def execute(self, reqStatus):
+        params = {}
+        url = '%s/data/info?request=%s' % (self.msConfig['msOutputUrl'],
+                                           wflow['RequestName'])
+        mgr = RequestHandler()
```
How about initializing it in the `__init__` method?

todor-ivanov replied:
Done

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +116,341 @@ def execute(self, reqStatus):
+        mgr = RequestHandler()
+        res = mgr.getdata(url, params=params, headers=headers, ckey=ckey(), cert=cert())
+        data = json.loads(res)['result'][0]
+        TransferInfo = data['transferDoc']
```
wrong variable name.

todor-ivanov replied:
Which one?

amaltaro replied:
`TransferInfo`. They should start with lower case ;)

todor-ivanov replied:
got it

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +116,341 @@ def execute(self, reqStatus):
+        # Find all the output placement rules created by the agents
+        # NOTE: Check MSOutput 'TransferStatus' first and assure it is 'done'
+        #       only then we proceed with building the respective list of rules
+        if wflow['TransferDone']:
```
Isn't rejected and aborted-completed workflows have this always set to its default value (False)? If so, then you know what will happen ;)

todor-ivanov replied:
Ok, I was trying to avoid some extra calls to Rucio, but yes... since we decided that  we will for sure try to clean those workflows too, we cannot avoid calling Rucio every time we meet one of them. So I've put them in a separated section in the `_dispatcher()` logic, and now anybody who calls `getRucioRules()` will receive what one was calling for regardless of the Transfer status in `MSOutput`.

todor-ivanov replied:
I just let them all (`['rejected', 'aborted-completed']`) pass through the pipeline and we get all those errors now: [1]. I am trying to figure out what should be done in this situation. 

[1]
```
2020-11-09 15:52:47,030:DEBUG:lock:_enter_create(): value creation lock <dogpile.cache.region._LockWrapper object at 0x7fa930185950> acquired
2020-11-09 15:52:47,031:DEBUG:lock:_enter_create(): Calling creation function for previously expired value
2020-11-09 15:52:47,031:DEBUG:lock:_enter_create(): Released creation lock
2020-11-09 15:52:47,034:DEBUG:connectionpool:_new_conn(): Starting new HTTP connection (1): cmsrucio-int.cern.ch:80
2020-11-09 15:52:47,087:DEBUG:connectionpool:_make_request(): http://cmsrucio-int.cern.ch:80 "GET /rules/?scope=cms&account=wma_test&name=%2FDYJetsToLL_Pt-50To100_TuneCUETP8M1_13TeV-amcatnloFXFX-pythia8%2FInteg_TestStep1-GENSIM_StepChain_Tasks_HG2011_Val_Todor_v1-v20%2FGEN-SIM HTTP/1.1" 200 None
2020-11-09 15:52:47,202:DEBUG:connectionpool:_make_request(): http://cmsrucio-int.cern.ch:80 "GET /rules/?scope=cms&account=wma_test&name=%2FDYJetsToLL_Pt-50To100_TuneCUETP8M1_13TeV-amcatnloFXFX-pythia8%2FInteg_TestStep1-GENSIM_StepChain_Tasks_HG2011_Val_Todor_v1-v20%2FLHE HTTP/1.1" 200 0
2020-11-09 15:52:47,231:DEBUG:connectionpool:_make_request(): http://cmsrucio-int.cern.ch:80 "GET /rules/?scope=cms&account=wma_test&name=%2FDYJetsToLL_Pt-50To100_TuneCUETP8M1_13TeV-amcatnloFXFX-pythia8%2FInteg_TestStep2-DIGI_StepChain_Tasks_HG2011_Val_Todor_v1-v20%2FGEN-SIM-RAW HTTP/1.1" 200 0
2020-11-09 15:52:47,260:DEBUG:connectionpool:_make_request(): http://cmsrucio-int.cern.ch:80 "GET /rules/?scope=cms&account=wma_test&name=%2FDYJetsToLL_Pt-50To100_TuneCUETP8M1_13TeV-amcatnloFXFX-pythia8%2FInteg_TestStep3-RECO_StepChain_Tasks_HG2011_Val_Todor_v1-v20%2FAODSIM HTTP/1.1" 200 0
2020-11-09 15:52:47,320:DEBUG:connectionpool:_make_request(): http://cmsrucio-int.cern.ch:80 "GET /dids/cms/%2FDYJetsToLL_Pt-50To100_TuneCUETP8M1_13TeV-amcatnloFXFX-pythia8%2FInteg_TestStep1-GENSIM_StepChain_Tasks_HG2011_Val_Todor_v1-v20%2FGEN-SIM HTTP/1.1" 404 233
2020-11-09 15:52:47,341:ERROR:MSRuleCleaner:_dispatchWflow(): plineAgentBlock: General error from pipeline. Workflow: tivanov_StepChain_Tasks_HG2011_Val_201029_112731_6371. Error: 
<@========== WMException Start ==========@>
Exception Class: WMRucioException
Message: Data identifier not found in Rucio: /DYJetsToLL_Pt-50To100_TuneCUETP8M1_13TeV-amcatnloFXFX-pythia8/Integ_TestStep1-GENSIM_StepChain_Tasks_HG2011_Val_Todor_v1-v20/GEN-SIM. Error: Data identifier not found.
Details: Data identifier 'cms:/DYJetsToLL_Pt-50To100_TuneCUETP8M1_13TeV-amcatnloFXFX-pythia8/Integ_TestStep1-GENSIM_StepChain_Tasks_HG2011_Val_Todor_v1-v20/GEN-SIM' not found
	ModuleName : WMCore.Services.Rucio.Rucio
	MethodName : isContainer
	ClassInstance : None
	FileName : /afs/cern.ch/user/t/tivanov/work/private/WMCoreDev.d/WMCore/src/python/WMCore/Services/Rucio/Rucio.py
	ClassName : None
	LineNumber : 759
	ErrorNr : 0

Traceback: 
  File "/afs/cern.ch/user/t/tivanov/work/private/WMCoreDev.d/WMCore/src/python/WMCore/Services/Rucio/Rucio.py", line 756, in isContainer
    response = self.cli.get_did(scope=scope, name=didName)

  File "/afs/cern.ch/work/t/tivanov/private/WMCoreDev.d/WMCore.venv/lib/python2.7/site-packages/rucio/client/didclient.py", line 394, in get_did
    raise exc_cls(exc_msg)

<@---------- WMException End ----------@>. 
Will retry again in the next cycle.
Traceback (most recent call last):
  File "/afs/cern.ch/user/t/tivanov/work/private/WMCoreDev.d/WMCore/src/python/WMCore/MicroService/Unified/MSRuleCleaner.py", line 218, in _dispatchWflow
    pline.run(wflow)
  File "/afs/cern.ch/user/t/tivanov/work/private/WMCoreDev.d/WMCore/src/python/Utils/Pipeline.py", line 140, in run
    return reduce(lambda obj, functor: functor(obj), self.funcLine, obj)
  File "/afs/cern.ch/user/t/tivanov/work/private/WMCoreDev.d/WMCore/src/python/Utils/Pipeline.py", line 140, in <lambda>
    return reduce(lambda obj, functor: functor(obj), self.funcLine, obj)
  File "/afs/cern.ch/user/t/tivanov/work/private/WMCoreDev.d/WMCore/src/python/Utils/Pipeline.py", line 72, in __call__
    return self.run(obj)
  File "/afs/cern.ch/user/t/tivanov/work/private/WMCoreDev.d/WMCore/src/python/Utils/Pipeline.py", line 75, in run
    return self.func(obj, *self.args, **self.kwargs)
  File "/afs/cern.ch/user/t/tivanov/work/private/WMCoreDev.d/WMCore/src/python/WMCore/MicroService/Unified/MSRuleCleaner.py", line 393, in getRucioRules
    blocks = self.rucio.getBlocksInContainer(dataCont)
  File "/afs/cern.ch/user/t/tivanov/work/private/WMCoreDev.d/WMCore/src/python/WMCore/Services/Rucio/Rucio.py", line 218, in getBlocksInContainer
    if not self.isContainer(container):
  File "/afs/cern.ch/user/t/tivanov/work/private/WMCoreDev.d/WMCore/src/python/WMCore/Services/Rucio/Rucio.py", line 759, in isContainer
    raise WMRucioException(msg)
WMRucioException: <@========== WMException Start ==========@>
Exception Class: WMRucioException
Message: Data identifier not found in Rucio: /DYJetsToLL_Pt-50To100_TuneCUETP8M1_13TeV-amcatnloFXFX-pythia8/Integ_TestStep1-GENSIM_StepChain_Tasks_HG2011_Val_Todor_v1-v20/GEN-SIM. Error: Data identifier not found.
Details: Data identifier 'cms:/DYJetsToLL_Pt-50To100_TuneCUETP8M1_13TeV-amcatnloFXFX-pythia8/Integ_TestStep1-GENSIM_StepChain_Tasks_HG2011_Val_Todor_v1-v20/GEN-SIM' not found
	ModuleName : WMCore.Services.Rucio.Rucio
	MethodName : isContainer
	ClassInstance : None
	FileName : /afs/cern.ch/user/t/tivanov/work/private/WMCoreDev.d/WMCore/src/python/WMCore/Services/Rucio/Rucio.py
	ClassName : None
	LineNumber : 759
	ErrorNr : 0

Traceback: 
  File "/afs/cern.ch/user/t/tivanov/work/private/WMCoreDev.d/WMCore/src/python/WMCore/Services/Rucio/Rucio.py", line 756, in isContainer
    response = self.cli.get_did(scope=scope, name=didName)

  File "/afs/cern.ch/work/t/tivanov/private/WMCoreDev.d/WMCore.venv/lib/python2.7/site-packages/rucio/client/didclient.py", line 394, in get_did
    raise exc_cls(exc_msg)

<@---------- WMException End ----------@>
2020-11-09 15:52:47,362:ERROR:MSRuleCleaner:_dispatchWflow(): plineArchive: Archival Error: MSRuleCleanerArchival: Not properly cleaned workflow: tivanov_StepChain_Tasks_HG2011_Val_201029_112731_6371.  Will retry again in the next cycle.
2020-11-09 15:52:47,363:DEBUG:MSRuleCleaner:_execute(): 
```

amaltaro replied:
Can't we catch a `WMRucioException` exception and simply pass it (well, log and pass)? See:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/Services/Rucio/Rucio.py#L794

In short, if DID hasn't been found, then we assume no data has been injected into Rucio for that container and there is nothing to be cleaned up.

todor-ivanov replied:
I agree the way out of this is to catch the exception as early as we make the request to Rucio for building these lists of clocks. What I do not like in the current situation (and something I try to work around right now) is this silent Exception metamorphosis that happens in the Rucio module [1], from a very well defined, clear and dedicated exception `DataIdentifierNotFound` which is unambiguously identifying the type of error we get from Rucio, to a more generic `WMRucioException` , in which the only clue that the did was not found is hidden inside the error message. And I do not want to deal with string parsing in order to distinguish exception types.


[1]
https://github.com/dmwm/WMCore/blob/51e3e2f6975b92abbbd9bf7291cf7e8d8eabd2e3/src/python/WMCore/Services/Rucio/Rucio.py#L794

amaltaro replied:
The idea with that generic WMCore exception was that clients of the Rucio module would not need to import all the possible exceptions that can happen from rucio.common. But I agree that having only one generic exception isn't ideal anymore.
How about you create another exception like `WMRucioDIDNotFound` and raise that one instead? If you do so, this needs to be communicated to the CRAB team though to make sure they are aware of this change.

todor-ivanov replied:
Thanks @amaltaro , tthat will fit my needs  perfectly. I will communicate the changes with Stefano, once I am ready with the so proposed Exception. 

Hi Stefano, @belforte, including you in this comment here, just for a heads up about an upcoming change.

todor-ivanov replied:
Hi Stefano @belforte, here is the change already prepared to be merged - these are the relevant commit and line [1] we are about to change, that may affect CRAB too. We are changing the more general `WMRucioException` we are rising from the `isContainer()` method in the `Rucio` module with a more specialized one, named `WMRucioDIDNotFoundException`. Here bellow  [2], in order to make your checks a little bit easier, I paste the full call tree and the exception generation/handling points in the code for both the `isContainer()` method and the `WMRucioException` itslef. In short if you export the `Rucio` module from WMcore inside CRAB or any of the methods listed bellow you should expect a change of `Exception` raising behavior. But if you do not catch the `WMRucioException` as  it is anywhere, then we shouldn't break the CRAB code with the current change. Hope this message helps to propagate the information correctly.
   


[1]
https://github.com/dmwm/WMCore/pull/10023/commits/2cda2ed9b64c0956c3fce1b36dba9e972bfa3f54#diff-2c120001070c945d3d9b54bda61f1707d86247a1306b2da392374ed611b62716R766

[2]
```
WMRucioException handling:

   raised from:
      WMCore/Services/Rucio/Rucio.py:
      listDataRules:
         from a general exception handling
      evaluateRSEExpression:
         from a InvalidRSEExpression Rucio exception handling
      isContainer:
         from a DataIdentifierNotFound Rucio exception handling
      createReplicationRule:
         from an AccessDenied Rucio exception handling
      getDataLockedAndAvailable_old:
         from an `if` statement
      getParentContainerRules:
         from an `if` statement
      getDataLockedAndAvailable_old:
         from an `if` statement

   handled at:
      WMComponent/RucioInjector/RucioInjectorPoller.py:_insertContainerRulesT0

Call Graph:
   WMCore/Services/Rucio/Rucio.py:
   isContainer:
      getDataLockedAndAvailable_old:
         TO BE DEPRECATED
      getBlocksInContainer:
         getReplicaInfoForBlocks:
            NOT CALLED
         _getContainerLockedAndAvailable_old:
            TO BE DEPRECATED
         getPileupLockedAndAvailable:
             src/python/WMCore/WMSpec/Steps/Fetchers/PileupFetcher.py:_getDatasetLocation
         getContainerLockedAndAvailable:
            NOT CALLED
      getPileupLockedAndAvailable:
         WMCore/WMSpec/Steps/Fetchers/PileupFetcher.py:_getDatasetLocation
      getDataLockedAndAvailable:
         WMCore/WorkQueue/DataLocationMapper.py:locationsFromRucio
         WMCore/WorkQueue/WorkQueue.py:_blockLocationRucioPhedex
         WMCore/WorkQueue/Policy/Start/StartPolicyInterface.py:getDatasetLocations
```

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +116,341 @@ def execute(self, reqStatus):
+        # NOTE: The current functional pipeline is always the last one in the PlineMarkers list
+        currPline = wflow['PlineMarkers'][-1]
+        delResutls = []
+        if wflow['TransferDone']:
```
Same comment as made on the function above. Some workflows will have `TransferDone=False`, thus not triggering rule deletion.

todor-ivanov replied:
Changed - The `TransferDone` flag is not checked  in the `cleanRucioRules()` method any more. We are about to guaranty only the proper workflows enter a cleaning pipeline.

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +116,341 @@ def execute(self, reqStatus):
+
+        # NOTE: The current functional pipeline is always the last one in the PlineMarkers list
+        currPline = wflow['PlineMarkers'][-1]
+        delResutls = []
```
typo on this variable name

todor-ivanov replied:
Fixed

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +116,341 @@ def execute(self, reqStatus):
+            delResutls.append(False)
+            msg = "%s: 'TransferStatus' is 'pending' or 'TransferInfo' is missing. "
+            msg += "Skip deleting any rules for: %s "
+            self.logger.debug(msg, currPline, wflow['RequestName'])
```
I think this log level can be made INFO.

todor-ivanov replied:
This message does not exist any more.

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +116,341 @@ def execute(self, reqStatus):
+        # ----------------------------------------------------------------------
+        # FIXME : To be removed once the plineMSTrBlock && plineMSTrCont are
+        #         developed
+        if wflow['CleanupStatus'][currPline] == 'plineMSTrBlock' \
```
```suggestion
        if wflow['CleanupStatus'][currPline] in ['plineMSTrBlock', 'plineMSTrCont']:
```

todor-ivanov replied:
This will eventually go away, but anyway, a good suggestion. 
Done.

amaltaro reviewed (changes requested):
Todor, please find in the code the second wave of reviews. Some are for future reference, some are minor, and others need to be considered. Thanks

todor-ivanov commented on src/python/WMCore/MicroService/DataStructs/MSRuleCleanerWorkflow.py:
```diff
@@ -0,0 +1,80 @@
+        myDoc = {}
+        for tup in self.docSchema():
+            if tup[0] in doc:
+                myDoc[tup[0]] = deepcopy(doc[tup[0]])
```
I know, but decided to stay safe than sorry :)

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +116,341 @@ def execute(self, reqStatus):
-        self.updateReportDict(summary, "error", 42)
+        # Call _execute() and feed the relevant pipeline with the objects popped from requestRecords
+        try:
+            total_num_requests, clean_num_requests, archived_num_requests = self._execute(requestRecords)
```
ok

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +116,341 @@ def execute(self, reqStatus):
+        #       This way we assure ourselves that we archive only workflows
+        #       that have accomplished the needed cleanup
+
+        clean_num_requests = 0
```
Done

todor-ivanov commented on src/python/WMCore/MicroService/DataStructs/MSRuleCleanerWorkflow.py:
```diff
@@ -0,0 +1,80 @@
+from copy import deepcopy
+
+
+class MSRuleCleanerWflow(dict):
```
Done for renaming the file.
Working on the unittests

todor-ivanov replied:
done

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -34,14 +54,60 @@ def __init__(self, msConfig, logger=None):
+        # Building the different set of plines we will need later:
+        # NOTE: The following are all the functional pipelines which are supposed to include
+        #       a cleanup function and report cleanup status in the MSRuleCleanerWflow object
+        self.cleanuplines = [self.plineMSTrCont,
```
I am not quite sure how exactly that  would work. Because am having those as object bound variables not because I access them from different pipelines from the same instance, but rather to use them as predefined lists/sest of different pipelines, which are not accessed from the `__init__()` but from a method down in the code. And I on the other hand did not want to recreate those lists every time a workflow enters the `_dispatcher()` ( which would be the only way to accomplish what you are suggesting here - IIUC), but rather keep those in the instance and just refer to the proper set/list of pipelines based on the criteria applied on the workflow inside the `_dispatcher()`. Or if I got you wrong, Please correct me.

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +116,341 @@ def execute(self, reqStatus):
+        self.currThreadIdent = self.currThread.name
+        self.updateReportDict(summary, "thread_id", self.currThreadIdent)
+
+        msg = "MSRuleCleaner is running in mode: %s."
```
Done

todor-ivanov committed:
Add WMRucioDIDNotFoundException.

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +121,337 @@ def execute(self, reqStatus):
+            msg += " Will retry again in the next cycle."
+            self.logger.info(msg, wflow['RequestName'])
+        elif wflow['RequestStatus'] == 'announced' and wflow['ParentageResolved'] and wflow['TransferDone']:
+            for pline in self.cleanuplines:
```
Unless I'm missing something, I don't think this code will ever be reached. We do not run the agent pipelines `self.agentlines` for workflows in `announced` status, which means we will never have `TransferDone=True`. Am I wrong?

todor-ivanov replied:
Sorry @amaltaro I will need you to clarify more on this, I did not get your comment here. This piece of code is executed every now and then, I do not understand two aspeccts of this comment, so I can clarify more, but :
1. What would `self.agentlines` do with the any if the conditions determining where we throw the current workflow - this variable is just a set of plines we throw at the relevant workflows nothing depends on it?
2. Why do you think we  would never have `TransferDone == True`

todor-ivanov replied:
Ok I think I may see your point .... but I think we may have locked out selves when we moved around some of the conditions under which we run the pipelines.

todor-ivanov replied:
Fixed

amaltaro reviewed (commented):
I have made one important comment on the code that needs clarification. Other than that (and the unit tests you're working on), everything looks fine!

todor-ivanov committed:
Fixing a lock condition from consulting a yet unresolved flag in _dispatcher()

Fixing a lock condition from consulting a yet unresolved flag in _dispatcher()

Revert "Fixing a lock condition from consulting a yet unresolved flag in _dispatcher()"

This reverts commit e17e92114938218c68f43bc95561ef07bd8457fa.

Early resolution of the 'TransferDone' flag.

Substitute list search with exact match.

UnitTests

Bugfix - enclose MSOutput REST call in a try statement && Typo.

loosen an if constraint.

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +119,343 @@ def execute(self, reqStatus):
+        a sequentially
+        """
+        # msg = "Called with wflow: %s"
+        # self.logger.debug(msg, pformat(wflow))
```
How about using this log, just to say which workflow is getting into the pipeline, something like:
```
self.loger.info("Dispatching workflow: %s", wflow['RequestName'])
```

or `Working on workflow: ....`

todor-ivanov replied:
that's a good idea ... I was using already using it in a similar way for some investigation, but forgot to remove it
Done now

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +119,343 @@ def execute(self, reqStatus):
+        params = {}
+        url = '%s/data/info?request=%s' % (self.msConfig['msOutputUrl'],
+                                           wflow['RequestName'])
+        # mgr = RequestHandler()
```
This comment can be removed. Question here: should we encapsulate this call in a try/exception block?

todor-ivanov replied:
You are absolutely correct. Before we were counting on the General exception handling  that was happening from the `pline.run()`, now since this one is at the top of the `_dispatcher()` things have changed a bit and if we break this call will break the whole run of the component too. Fixing it now.

todor-ivanov replied:
Done

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +119,343 @@ def execute(self, reqStatus):
+        res = self.curlMgr.getdata(url, params=params, headers=headers, ckey=ckey(), cert=cert())
+        data = json.loads(res)['result'][0]
+        transferInfo = data['transferDoc']
+        if transferInfo is not None and TransferInfo['TransferStatus'] == 'done':
```
bug/typo here

todor-ivanov replied:
nice catch Thanks

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -50,27 +119,343 @@ def execute(self, reqStatus):
+            msg = "Skipping workflow: %s - 'TransferStatus' is 'pending' or 'TransferInfo' is missing in MSOutput."
+            msg += " Will retry again in the next cycle."
+            self.logger.info(msg, wflow['RequestName'])
+        elif wflow['RequestStatus'] == 'announced' and wflow['ParentageResolved'] and wflow['TransferDone']:
```
Only this:
```
elif wflow['RequestStatus'] == 'announced':
```
would do!

todor-ivanov replied:
Allow me to disagree on this one. The very strict definition on which workflows to work in this line is by purpose. And it is because if we generalize the condition here, alot of things that have not been thought on would pass under the generalized condition, and in case there is an error in the logic which workflows reach to here (as an example someone decides to tamper with or by mistake touches the list of statuses for the component run) then all those will pass through unnoticed and unpredictable behavior is guaranteed. But on the other hand if we see something falling to the `else` statement a line bellow this is going to be a sure signal of something wrong, wile protecting us from screwing a workflow.

todor-ivanov replied:
What I am trying to say is basically this one is there to protect ourselves from our won mistakes, and for me it makes it much clearer and visibe which is the full set of conditions about what workflow Ii expect here.

amaltaro replied:
And those would still be protected with the suggestion I made. Announced workflows have to satisfy only 2 conditions:
* ParentageResolved
* TransferDone
which are already evaluated in the previous `elif`. I'm also not suggesting you to remove the `else`.
My suggetion makes the code cleaner and avoid repetitive/duplicate check conditions.

todor-ivanov replied:
You have it.

amaltaro reviewed (changes requested):
Todor, I have made one standalone comment in the code and a few more that will get published with this review. A couple of them must be looked into.
In addition to that, one of the Rucio unit tests now fail with the new exception put in place. Please look at the jenkins report.

todor-ivanov committed:
UnitTests

UnitTest: MSRuleCleaner_t

UnitTest: MSRuleCleaner_t: Add pipelines tests && MSRuleCleanerWflow_t.py: Read reqest from file.

UnitTest: MSRuleCleaner_t: Add _checkClean test.

UnitTest: MSRuleCleaner_t: _checkClean: Add tests for skipped and unclean wflows archival.

amaltaro commented on test/python/WMCore_t/MicroService_t/Unified_t/MSRuleCleaner_t.py:
```diff
@@ -0,0 +1,197 @@
+        self.assertFalse(self.msRuleCleaner._checkClean(wflowFlags))
+
+        # 2. MaskList Empty
+        wflowFlags = {'CleanupStatus': {'plineAgentBlock': True, 'plineAgentCont': True},
```
Just out of curiosity, is it a valid case? I mean, can we have a pipeline name in the CleanupStatus dict but not marked in the PlineMarkers?

todor-ivanov replied:
No, The code as  it is right now should have other protections against this case, but I put this check the unit test for completeness, because the `_checkClean()` function mast cover all the 4 types  of combinations regardless if the rest of the code has an additional protection or not. In this case - the `cleanRucioRules()` function is reading the `currentPline` from the `PlineMarkers` list, but in case we make a change about this, I think this should have been thought of in advance in the `_checkClean()` and we assure the code is doing the right thing even if we try to break it.

amaltaro reviewed (approved):
Thanks Todor, everything looks good now.
I didn't see the k8s configuration for this new microservice, haven't you created them yet? We need those to move forward...

todor-ivanov commented:
Working on the K8 configs in this very moment.

amaltaro commented:
Thanks Todor.
